tests/functional_simulator/exe_stage_tests.cpp

```diff
@@ -0,0 +1,244 @@
+
+    // Verify the ALU result is correct (5 * 7 = 35)
+    const PipelineStageData* execute_data = getExecuteStageData();
+    EXPECT_EQ(execute_data->alu_result, 35);
+    EXPECT_TRUE(execute_data->dest_reg.has_value());
+    EXPECT_EQ(execute_data->dest_reg.value(), 3);
+}
+
+// Test AND instruction execution
+TEST_F(ExecuteStageTest, ExecuteAND) {
+    // Set up execute stage with AND instruction and test values
+    setupExecuteStage(AND_INSTR, 0b1010, 0b1100, 3);  // rs=10, rt=12, dest=3
+
+    // Execute the instruction
+    sim->execute();
+
+    // Verify the ALU result is correct (0b1010 & 0b1100 = 0b1000 = 8)
+    const PipelineStageData* execute_data = getExecuteStageData();
+    EXPECT_EQ(execute_data->alu_result, 8);
+    EXPECT_TRUE(execute_data->dest_reg.has_value());
+    EXPECT_EQ(execute_data->dest_reg.value(), 3);
+}
+
+// Test BEQ instruction execution with branch taken
+TEST_F(ExecuteStageTest, ExecuteBEQTaken) {
+    // Set up execute stage with BEQ instruction and equal values
+    setupExecuteStage(BEQ_INSTR, 25, 25, std::nullopt, 1000);  // rs=25, rt=25, no dest, pc=1000
+
+    // Execute the instruction
+    sim->execute();
+
+    // Verify branch was taken and target address calculated correctly
+    const PipelineStageData* execute_data = getExecuteStageData();
+    EXPECT_EQ(execute_data->alu_result, 1200);  // 1000 + (50*4) = 1200
+    EXPECT_TRUE(sim->isBranchTaken());          // Use getter method
+    EXPECT_FALSE(execute_data->dest_reg.has_value());  // No writeback for branches
+}
+
+// Test BEQ instruction execution with branch not taken
+TEST_F(ExecuteStageTest, ExecuteBEQNotTaken) {
+    // Set up execute stage with BEQ instruction and unequal values
+    setupExecuteStage(BEQ_INSTR, 25, 30, std::nullopt, 1000);  // rs=25, rt=30, no dest, pc=1000
+
+    // Execute the instruction
+    sim->execute();
+
+    // Verify branch was not taken
+    const PipelineStageData* execute_data = getExecuteStageData();
+    EXPECT_FALSE(sim->isBranchTaken());         // Use getter method
+    EXPECT_FALSE(execute_data->dest_reg.has_value());
+}
+
+// Test BZ instruction with branch taken
+TEST_F(ExecuteStageTest, ExecuteBZTaken) {
+    // Set up execute stage with BZ instruction and zero value
+    setupExecuteStage(BZ_INSTR, 0, 50, std::nullopt, 1000);  // rs=0, rt=50 (offset), no dest, pc=1000
+
+    // Execute the instruction
+    sim->execute();
+
+    // Verify branch was taken (50*4 + 1000 = 1200)
+    const PipelineStageData* execute_data = getExecuteStageData();
+    EXPECT_EQ(execute_data->alu_result, 1200);
+    EXPECT_TRUE(sim->isBranchTaken());          // Use getter method
+    EXPECT_FALSE(execute_data->dest_reg.has_value());
+}
+
+// Test BZ instruction with branch not taken
+TEST_F(ExecuteStageTest, ExecuteBZNotTaken) {
+    // Set up execute stage with BZ instruction and non-zero value
+    setupExecuteStage(BZ_INSTR, 5, 50, std::nullopt, 1000);  // rs=5 (non-zero), rt=50, no dest, pc=1000
```

Comment:
Both of these BZ test comments are a bit misleading too, since BZ doesn't use Rt, the immediate value is used for the address calculation. The logic in the method seems great, but it might be worth not setting rt to the same value as the immediate value and making a note of how the calculation is performed.